File: sdks/python/src/ankka/effects/key_value.py
"""Effects for key value entities: replace the state, then decide what to reply."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import timedelta
from typing import Any, Generic, TypeVar

from ankka.context import Metadata
from ankka.effects.common import DeleteNow, Error, ErrorCode, ExpireAfter, Fail, NoReply, Outcome, Reply, Retention

S = TypeVar("S")
R = TypeVar("R")


@dataclass(frozen=True)
class KeyValueEffect(Generic[S, R]):
    new_state: S | None  # None: unchanged
    retention: Retention | None
    outcome: Outcome


@dataclass(frozen=True)
class KeyValueReadOnlyEffect(KeyValueEffect[S, R]):
    new_state: S | None = None
    retention: Retention | None = None
    outcome: Outcome = NoReply()


@dataclass(frozen=True)
class UpdateBuilder(Generic[S]):
    new_state: S | None
    retention: Retention | None = None

    def delete_entity(self) -> UpdateBuilder[S]:
        return UpdateBuilder(self.new_state, DeleteNow())

    def expire_after(self, duration: timedelta) -> UpdateBuilder[S]:
        return UpdateBuilder(self.new_state, ExpireAfter(duration))

    def then_reply(self, compute: Callable[[S], R], metadata: Metadata | None = None) -> KeyValueEffect[S, R]:
        return KeyValueEffect(self.new_state, self.retention, Reply(compute, metadata or Metadata()))

    def then_reply_state(self) -> KeyValueEffect[S, S]:
        return KeyValueEffect(self.new_state, self.retention, Reply(lambda s: s))

    def then_no_reply(self) -> KeyValueEffect[S, Any]:
        return KeyValueEffect(self.new_state, self.retention, NoReply())


class KeyValueEffects(Generic[S]):
    def update_state(self, state: S) -> UpdateBuilder[S]:
        return UpdateBuilder(state)

    def delete_entity(self) -> UpdateBuilder[S]:
        return UpdateBuilder(None, DeleteNow())

    def reply(self, value: R, metadata: Metadata | None = None) -> KeyValueReadOnlyEffect[S, R]:
        return KeyValueReadOnlyEffect(outcome=Reply(lambda _: value, metadata or Metadata()))

    def error(self, message: str, code: ErrorCode = ErrorCode.BAD_REQUEST) -> KeyValueReadOnlyEffect[S, Any]:
        return KeyValueReadOnlyEffect(outcome=Fail(Error(message, code)))

    def no_reply(self) -> KeyValueReadOnlyEffect[S, Any]:
        return KeyValueReadOnlyEffect(outcome=NoReply())

File: sdks/typescript/src/effects/keyValue.ts
// Effects for key value entities: replace the state, then decide what to reply.

import type { Duration } from "../time.ts"
import { done, type Done } from "../schema.ts"
import { ErrorCode, Outcome, Retention, type EffectLike, type Metadata, type ReadOnlyLike } from "./common.ts"

export interface UpdateEffect<S, R> extends EffectLike<R> {
  readonly kind: "update"
  /** The state to store; `null` when the effect only deletes. */
  readonly newState: S | null
  readonly retention: Retention | null
  readonly outcome: Outcome<S, R>
}

export interface KeyValueReadOnlyEffect<S, R> extends ReadOnlyLike<R> {
  readonly kind: "read-only"
  readonly retention: null
  readonly outcome: Outcome<S, R>
}

export type KeyValueEffect<S, R> = UpdateEffect<S, R> | KeyValueReadOnlyEffect<S, R>

export class UpdateBuilder<S> {
  readonly #newState: S | null
  readonly #retention: Retention | null

  constructor(newState: S | null, retention: Retention | null = null) {
    this.#newState = newState
    this.#retention = retention
  }

  deleteEntity(): UpdateBuilder<S> {
    return new UpdateBuilder(this.#newState, Retention.deleteNow)
  }

  expireAfter(after: Duration): UpdateBuilder<S> {
    return new UpdateBuilder(this.#newState, Retention.expireAfter(after))
  }

  /** Reply with a value computed from the new state. */
  thenReply<R>(compute: (state: S) => R, metadata?: Metadata): UpdateEffect<S, R> {
    return Object.freeze({ kind: "update", newState: this.#newState, retention: this.#retention, outcome: Outcome.reply(compute, metadata) })
  }

  thenReplyState(): UpdateEffect<S, S> {
    return this.thenReply((state) => state)
  }

  thenNoReply(): UpdateEffect<S, never> {
    return Object.freeze({ kind: "update", newState: this.#newState, retention: this.#retention, outcome: Outcome.noReply() })
  }
}

/** The factory a key value entity's handler reaches as `this.effects`. */
export class KeyValueEffects<S> {
  updateState(state: S): UpdateBuilder<S> {
    return new UpdateBuilder<S>(state)
  }

  /** Delete the entity, keeping the state as it is until then. */
  deleteEntity(): UpdateBuilder<S> {
    return new UpdateBuilder<S>(null, Retention.deleteNow)
  }

  reply<R>(value: R, metadata?: Metadata): KeyValueReadOnlyEffect<S, R> {
    return Object.freeze({ kind: "read-only", retention: null, outcome: Outcome.reply(() => value, metadata) })
  }

  error(message: string, code: ErrorCode = ErrorCode.BadRequest): KeyValueReadOnlyEffect<S, never> {
    return Object.freeze({ kind: "read-only", retention: null, outcome: Outcome.fail(message, code) })
  }

  noReply(): KeyValueReadOnlyEffect<S, never> {
    return Object.freeze({ kind: "read-only", retention: null, outcome: Outcome.noReply() })
  }
}

export { done, type Done }
